bindings/node/lib/stateMachine.js

```diff
@@ -0,0 +1,231 @@
+    switch (context.state) {
+      case MONGOCRYPT_CTX_NEED_MONGO_COLLINFO: {
+        const filter = bson.deserialize(context.nextMongoOperation());
+        this.fetchCollectionInfo(client, context.ns, filter, (err, collInfo) => {
+          if (err) {
+            return callback(err, null);
+          }
+
+          if (collInfo) {
+            context.addMongoOperationResponse(collInfo);
+          }
+
+          context.finishMongoOperation();
+          this.execute(autoEncrypter, context, callback);
+        });
+
+        return;
+      }
+
+      case MONGOCRYPT_CTX_NEED_MONGO_MARKINGS: {
+        const command = context.nextMongoOperation();
+        this.markCommand(mongocryptdClient, context.ns, command, (err, markedCommand) => {
+          if (err) return callback(err, null);
+          context.addMongoOperationResponse(markedCommand);
+          context.finishMongoOperation();
+
+          this.execute(autoEncrypter, context, callback);
+        });
+
+        return;
+      }
+
+      case MONGOCRYPT_CTX_NEED_MONGO_KEYS: {
+        const filter = context.nextMongoOperation();
+        this.fetchKeys(client, keyVaultNamespace, filter, (err, keys) => {
+          if (err) return callback(err, null);
+          keys.forEach(key => {
+            context.addMongoOperationResponse(bson.serialize(key));
+          });
+
+          context.finishMongoOperation();
+          this.execute(autoEncrypter, context, callback);
+        });
+
+        return;
+      }
+
+      case MONGOCRYPT_CTX_NEED_KMS: {
+        const promises = [];
+
+        let request;
+        while ((request = context.nextKMSRequest())) {
+          promises.push(this.kmsRequest(request));
+        }
+
+        Promise.all(promises)
+          .then(() => {
+            context.finishKMSRequests();
+            this.execute(autoEncrypter, context, callback);
+          })
+          .catch(err => {
+            callback(err, null);
+          });
+
+        return;
+      }
+
+      // terminal states
+      case MONGOCRYPT_CTX_READY:
+        callback(null, bson.deserialize(context.finalize()));
+        return;
+
+      case MONGOCRYPT_CTX_ERROR: {
+        const message = context.status.message;
+        callback(new MongoCryptError(message));
+        return;
+      }
+
+      case MONGOCRYPT_CTX_DONE:
+        return;
+
+      default:
+        callback(new MongoCryptError(`Unknown state: ${context.state}`));
+        return;
+    }
+  }
+
+  /**
+   *
+   * @param {*} kmsContext
+   */
+  kmsRequest(request) {
+    const options = { host: request.endpoint, port: 443 };
+    const message = request.message;
+
+    return new Promise((resolve, reject) => {
+      const socket = tls.connect(options, () => {
+        socket.write(message);
+      });
+
+      socket.once('timeout', () => {
+        socket.removeAllListeners();
+        socket.destroy();
+        reject(new Error('KMS request timed out'));
```

Comment:
Should we make this our `MongoCryptError`?

Reply:
done